### ISSUE TITLE
`Cache(writeOutputs, ...)` is not retrieving a the output raster obj

### ISSUE THREAD

**Issue by CeresBarros** [PredictiveEcology/reproducible#316]
I'm not sure what is going on, but after updating `reproducible` to use the head of the `development` branch I'm hitting two strange behaviours with `Cache(writeOutputs, ...)`.
1) the second call to `Cache(writeOutputs, ...` doesn't seem to retrive the first call's cached output.
2) the third call retrieves a cached output, but it's no longer a raster layer

```
Require::Require("PredictiveEcology/reproducible@development (HEAD)")

options("reproducible.cacheSaveFormat" = "qs",
        "reproducible.useNewDigestAlgorithm" = 2,
        "reproducible.useCache" = TRUE,
        "reproducible.useGDAL" = FALSE,
        "reproducible.useMemoise" = TRUE,
        "reproducible.useTerra" = TRUE,
        "reproducible.rasterRead" = "terra::rast")

dPath <- file.path(tempdir2())
cPath <- file.path(dPath, "cache") 
remoteTifUrl <- "https://github.com/rspatial/terra/raw/master/inst/ex/elev.tif"

ras <- prepInputs(url = remoteTifUrl, res = 250,
                  destinationPath = dPath,
                  cachePath = cPath)
ras2 <- Cache(writeOutputs, ras, 
             filename2 = .suffix(file.path(dPath, "ras.tif"), 
                                 paste0("_test")), datatype = "INT2U", 
             overwrite = TRUE, userTags = c("ras"), 
             omitArgs = c("userTags"),
             cachePath = cPath)
ras2

## this doens't seem to load the cached call. Does the call above change ras?
ras2 <- Cache(writeOutputs, ras, 
             filename2 = .suffix(file.path(dPath, "ras.tif"), 
                                 paste0("_test")), datatype = "INT2U", 
             overwrite = TRUE, userTags = c("ras"), 
             omitArgs = c("userTags"),
             cachePath = cPath)
## need to run again
ras2 <- Cache(writeOutputs, ras, 
              filename2 = .suffix(file.path(dPath, "ras.tif"), 
                                  paste0("_test")), datatype = "INT2U", 
              overwrite = TRUE, userTags = c("ras"), 
              omitArgs = c("userTags"),
              cachePath = cPath)

ras2
```

session info:
```
 sessionInfo()
R version 4.3.0 (2023-04-21 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default


locale:
[1] LC_COLLATE=English_Canada.utf8  LC_CTYPE=English_Canada.utf8    LC_MONETARY=English_Canada.utf8
[4] LC_NUMERIC=C                    LC_TIME=English_Canada.utf8    

time zone: America/Vancouver
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] reproducible_2.0.2.9001

loaded via a namespace (and not attached):
 [1] bit_4.0.5          Require_0.2.6.9008 dplyr_1.1.2        compiler_4.3.0     crayon_1.5.2       tidyselect_1.2.0  
 [7] Rcpp_1.0.10        blob_1.2.4         parallel_4.3.0     fastdigest_0.6-3   fastmap_1.1.1      R6_2.5.1          
[13] lobstr_1.1.2       generics_0.1.3     classInt_0.4-9     curl_5.0.0         sf_1.0-12          tibble_3.2.1      
[19] units_0.8-2        DBI_1.1.3          pillar_1.9.0       rlang_1.1.1        utf8_1.2.3         cachem_1.0.8      
[25] terra_1.7-29       fs_1.6.2           bit64_4.0.5        RSQLite_2.3.1      memoise_2.0.1      cli_3.6.1         
[31] magrittr_2.0.3     class_7.3-21       grid_4.3.0         digest_0.6.31      rstudioapi_0.14    lifecycle_1.0.3   
[37] vctrs_0.6.2        KernSmooth_2.23-20 proxy_0.4-27       fpCompare_0.2.4    gargle_1.4.0       glue_1.6.2        
[43] data.table_1.14.8  sessioninfo_1.2.2  codetools_0.2-19   e1071_1.7-13       googledrive_2.1.0  fansi_1.0.4       
[49] purrr_1.0.1        httr_1.4.6         tools_4.3.0        pkgconfig_2.0.3   
```

**Comment by CeresBarros** [comment]
After addressing #318 , it seems I now have a different error:
```
Error in returnObjFromRepo(isInRepo = isInRepo, notOlderThan = notOlderThan,  : 
  Error : [rast] file does not exist: D:/GitHub/reproducible/C:/Users/cbarros/AppData/Local/Temp/RtmpgXm2es/reproducible/BetWDxdH/rasterToMatchLarge_test.tif

Error in trying to recover cacheID: 1e252361e95e3178
You will likely need to remove that item from Cache, e.g., 
clearCache('C:/Users/cbarros/AppData/Local/Temp/RtmpgXm2es/reproducible/BetWDxdH/cache', userTags = '1e252361e95e3178')
In addition: Warning messages:
1: In FUN(X[[i]], ...) :
  cannot create dir 'D:\GitHub\reproducible\C:\Users', reason 'Invalid argument'
2: In file.create(to[okay]) :
  cannot create file 'D:/GitHub/reproducible/C:/Users/cbarros/AppData/Local/Temp/RtmpgXm2es/reproducible/BetWDxdH/rasterToMatchLarge_test.tif', reason 'Invalid argument'
3: `D:/GitHub/reproducible/C:/Users/cbarros/AppData/Local/Temp/RtmpgXm2es/reproducible/BetWDxdH/rasterToMatchLarge_test.tif' does not exist in the file system, and is not recognized as a supported dataset name. (GDAL error 4) 
```

```
Require::Require(c("CeresBarros/reproducible@82796fa4ee36c341ec455163ef84cb3a60860cad (HEAD)",
                   "terra"))

options("reproducible.rasterRead" = "terra::rast")
options("reproducible.useTerra" = TRUE)

dPath <- file.path(tempdir2())
cPath <- file.path(dPath, "cache") 

options("reproducible.cachePath" = cPath)
options("reproducible.destinationPath" = dPath)

cacheTags <- "test"
studyAreaName <- "test"
destinationPath <- dPath

f <- system.file("ex/elev.tif", package="terra")
rasterToMatchLarge <- rast(f)

RTMvals <- as.vector(rasterToMatchLarge[])
rasterToMatchLarge[!is.na(RTMvals)] <- 1

rasterToMatchLarge2 <- Cache(writeOutputs, rasterToMatchLarge, 
                             filename2 = .suffix(file.path(destinationPath, "rasterToMatchLarge.tif"), 
                                                 paste0("_", studyAreaName)), datatype = "INT2U", 
                             overwrite = TRUE, userTags = c(cacheTags, "rasterToMatchLarge"), 
                             omitArgs = c("userTags"))
rasterToMatchLarge2  ## all good on first call

## try again
rasterToMatchLarge2 <- Cache(writeOutputs, rasterToMatchLarge, 
                             filename2 = .suffix(file.path(destinationPath, "rasterToMatchLarge.tif"), 
                                                 paste0("_", studyAreaName)), datatype = "INT2U", 
                             overwrite = TRUE, userTags = c(cacheTags, "rasterToMatchLarge"), 
                             omitArgs = c("userTags"))
rasterToMatchLarge2  ## all good on second call

## try again -- error, can't find file
rasterToMatchLarge2 <- Cache(writeOutputs, rasterToMatchLarge, 
                             filename2 = .suffix(file.path(destinationPath, "rasterToMatchLarge.tif"), 
                                                 paste0("_", studyAreaName)), datatype = "INT2U", 
                             overwrite = TRUE, userTags = c(cacheTags, "rasterToMatchLarge"), 
                             omitArgs = c("userTags"))
```

**Comment by eliotmcintire** [comment]
It looks like there are 2 issues going on here:

1. In the case where the *second* `Cache` call does not recover a cached object, this is a case of needing to specify `filename2` as a `quick` argument. `Cache` tries to evaluate all arguments... the first time `writeOutputs` is called, the file identified as `filename2` doesn't exist yet, so the argument is assessed as a character string. The second time through, the `filename2` file does exist, so `Cache` assesses it as a file and fails to recover the cached copy because it is a different call. The third time through, it recovers the 2nd call.  This is not a bug, but an annoyance. See the `quick` section in `?Cache`. The call should look like this:

```
ras2 <- Cache(writeOutputs, ras, 
              filename2 = .suffix(file.path(dPath, "ras.tif"), 
                                  paste0("_test")), datatype = "INT2U", 
              quick = "filename2",   # <----------------------------------------------
              overwrite = TRUE, userTags = c("ras"), 
              omitArgs = c("userTags"),
              cachePath = cPath)
```

2. There is a bug for the `SpatRaster` class objects, when the function (in this case, `writeOutputs`) has an objective of changing the path of the `SpatRaster`. I have pushed a possible fix in PR #322 . Waiting for GitHub actions. So the code would now look like this, using `quick = "filename2"` and this now works for me:
```
Require::Require(c("PredictiveEcology/reproducible@d034c17787278976bd6c1f4b71698ee9fdc2d862 (HEAD)",
                   "terra"))

options("reproducible.rasterRead" = "terra::rast")
options("reproducible.useTerra" = TRUE)

dPath <- file.path(tempdir2())
cPath <- file.path(dPath, "cache") 

options("reproducible.cachePath" = cPath)
options("reproducible.destinationPath" = dPath)

cacheTags <- "test"
studyAreaName <- "test"
destinationPath <- dPath

f <- system.file("ex/elev.tif", package="terra")
rasterToMatchLarge <- rast(f)

RTMvals <- as.vector(rasterToMatchLarge[])
rasterToMatchLarge[!is.na(RTMvals)] <- 1

rasterToMatchLarge2 <- Cache(writeOutputs, rasterToMatchLarge, 
                             filename2 = .suffix(file.path(destinationPath, "rasterToMatchLarge.tif"), 
                                                 paste0("_", studyAreaName)), datatype = "INT2U", 
                             overwrite = TRUE, userTags = c(cacheTags, "rasterToMatchLarge"), 
                             quick = "filename2",
                             omitArgs = c("userTags"))

rasterToMatchLarge2 <- Cache(writeOutputs, rasterToMatchLarge, 
                             filename2 = .suffix(file.path(destinationPath, "rasterToMatchLarge.tif"), 
                                                 paste0("_", studyAreaName)), datatype = "INT2U", 
                             overwrite = TRUE, userTags = c(cacheTags, "rasterToMatchLarge"), 
                             quick = "filename2",
                             omitArgs = c("userTags"))

**Comment by CeresBarros** [comment]
Thanks @eliotmcintire.
